### DISCUSSION TITLE
Collegamento remoto ad Home Assistant con Reverse Proxy

### DISCUSSION THREAD

**Discussion by lnrd00** [indomus/forum#2863] [HUB: Home Assistant]
Salve, ho un dubbio per quanto riguarda il collegamento da remoto ad Home Assistant tramite Reverse Proxy.
Ho seguito la vostra guida e tutto funziona correttamente, ma periodicamente mi diventa impossibile accedere da remoto.
Andando sui logs di home assistant, noto che l’IP che mi appare in

```
ERROR (MainThread) [homeassistant.components.http.forwarded] Received X-Forwarded-For header from an untrusted proxy
```
**è un IP che col tempo è sempre diverso**, e dunque ogni volta devo aggiornare il file configuration.yaml nella sua sezione

```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - XXX.XXX.XXX.XXX
```

La variabilità dell’IP del proxy è dovuta ad un mio errore o è normale che faccia così? C’è un modo per rendere statico questo IP? Altrimenti mi capita di trovarmi fuori casa senza la possibilità di poter accedere ad Home Assistant.
P.S. Io ho Caddy installato con docker su Raspberry, insieme ad Home Assistant.

Grazie in anticipo!

**Comment by snakuzzo** [comment]
Ciao.
Quell'IP cambia/può cambiare ogni volta che ricrei il container. Il classico caso è quando aggiorni la versione di un'applicazione. In quel caso viene rimosso il container e ricreato con la nuova immagine caricata. Di default docker ha una sua rete e sceglie lui l'IP da assegnare al container nuovo (sempre se non definito diversamente).

Al posto del singolo IP, ti conviene indicare tutta la subnet di docker (come indicato [QUI](https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reverse-proxy/#homeassistant)]

Se metti `172.0.0.0/8` dovresti essere a posto per sempre.

Non so se usi docker compose o meno, ma nel caso ti consiglio di utilizzarlo e di definire staticamente una tua network. Ti rimando alla guida di compose sulle network [QUI](https://docs.docker.com/compose/compose-file/06-networks/)

Piccolo spoiler...

```yaml
networks:
  domotica-compose:
    name: domotica-compose
    ipam:
      driver: default
      config:
        - subnet: "172.29.1.0/27"
          gateway: "172.29.1.1"
```

questo è come ho configurato io la rete. Nel mio caso nel `configuration.yaml` di Core ho inserito `172.29.1.0/27` che è la subnet della network di tutti i container (tra cui Caddy) con configurazione di rete `bridge` che fanno parte dello stesso compose

**Reply by lnrd00** [reply]
Ti ringrazio!
Per essere più preciso, gli IP del reverse proxy che mi sono stati assegnati erano del tipo `172.18.0.2` oppure `172.18.0.4`, ciò significa che per indicare tutta la subnet dovrei mettere `172.18.0.0/8` oppure va bene `172.0.0.0/8` ?

Ancora un'altra domanda: mi sto rendendo conto che in realtà nei logs non appare più l'errore
```
ERROR (MainThread) [homeassistant.components.http.forwarded] Received X-Forwarded-For header from an untrusted proxy
```
Ovvero, pur non riuscendo ad accedere dall'esterno, i logs di homeassistant non mi segnalano questo problema, dunque come se il problema fosse un altro.
Cosa potrebbe essere?

**Reply by snakuzzo** [reply]
Per il discorso IP...la tua osservazione è sbagliata.
`172.0.0.0/8` include tutti gli IP da `172.0.0.0` a `172.255.255.255`, incluso quello che vedi tu. Scrivere la subnet come hai fatto tu `172.18.0.0/8` non ha tanto senso, oltre a rappresentare lo stesso range.
Non fissarti sul `172.18` perché creando una nuova network Docker proprio quello potrebbe cambiare e diventare `172.19`...quindi lascia quello che ti ho detto.

Sul perché continui a non accedere...se ti limiti a scrivere quello che hai appena scritto, solo una sfera di cristallo magica potrebbe dirci il motivo. Devi descrivere meglio qual è il problema. Che evidenze hai?

**Reply by aventrax** [reply]
In realtà la rete privata di classe b è 172.16.0.0/12, non c'è bisogno di usare la /8 che fra l'altro include anche ip pubblici.

**Reply by snakuzzo** [reply]
sì, giusto...è una subnet troppo ampia. ma non è detto che sia `172.18`. Se smanetti e reinstalli docker o fai altri magheggi, potrebbe avere anche una classe diversa dalla (es. `172.19`) ... e comunque non ha assolutamente importanza.
Quello che sarebbe giusto fare, per avere una soluzione sempre sicuramente funzionante, è creare la network indicando esplicitamente il CIDR (come indicato prima).

**Reply by lnrd00** [reply]
Grazie per le informazioni. Per quanto riguarda la network, sto effettivamente utilizzando Docker Compose e quindi sto provando a crearne una. Non sapevo esattamente dove trovare la subnet e il gateway di tutti i container, dunque ho provato con ```docker network inspect bridge```. Va bene? Ciò che mi rende è questo 
```
[
    {
        "Name": "bridge",
        "Id": "6b443b84af235242d9fa83458cbb15b388bbfa7b5db1ad30821c353b004f668e",
        "Created": "2024-01-23T15:25:41.587619383+01:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500"
        },
        "Labels": {}
    }
]
```
Dove immagino che ` "Subnet": "172.17.0.0/16" e "Gateway": "172.17.0.1"` sia ciò che interessa a me.

Se quello che ho detto finora è corretto, avrei anche un'altra domanda: ciò che mi hai mostrato della configurazione della tua rete è tutto codice che hai scritto all'interno del `configuration.yaml` di homeassistant?

**Reply by lnrd00** [reply]
Per quanto invece riguarda il "malfunzionamento" dell'accesso da remoto, adesso se provo ad accedere da remoto con ```https://ilmiodominio.duckdns.org:9000``` semplicemente la pagina non carica, ne da browser né dall'app, e in quest'ultima rende un messaggio di errore allegato sotto dopo diversi minuti, che sembrerebbe indicare un errore dovuto ad un timeout:

![IMG_5600](https://github.com/indomus/forum/assets/154708323/71bfc748-3d75-4fd0-97d5-048eff5f2025)

Ho notato il problema da circa una settimana, ma non sono intervenuto subito anche perché pensavo fosse l'ennesimo problema dovuto ad una modifica dell'IP del proxy, invece sembrerebbe non essere così dato che:

1.  Ho aggiornato 
```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - XXX.XXX.XXX.XXX
```
inserendo `172.0.0.0/8` come consigliato.

2. Sui logs di Home Assistant non appare l'errore ```ERROR (MainThread) [homeassistant.components.http.forwarded] Received X-Forwarded-For header from an untrusted proxy``` come appariva in caso di variazione dell'IP del proxy.

Quindi sì, il motivo con cui ho aperto la discussione potrebbe non essere più pertinente e adesso i miei obbiettivi principali sono:

- Comprendere come creare una nuova network docker
- comprendere l'origine di questo problema "misterioso" per l'accesso remoto

Grazie in anticipo!!

**Reply by snakuzzo** [reply]
Dovresti prima fare un po' d'ordine.
Se usi `docker compose` e non definisci esplicitamente la network (incluso il nome), lui ti creerà in automatico una network ad hoc per quel compose, che avrà una classe di indirizzamento diversa dalla network `bridge` (che è quella che viene creata di default da docker).
Guarda... ammettiamo tu abbia una directory `domotica` con all'interno il tuo compose con la definizione di tutti i tuoi servizi...guardando la lista di network avresti qualcosa del genere...

Es.
```
$ docker network ls
NETWORK ID     NAME                           DRIVER    SCOPE
0e85b36e8b8b   bridge                         bridge    local
39fdf9ca14af   domotica_default               bridge    local
d2acfa2415c2   host                           host      local
846178b1790d   main                           bridge    local
```

Per recuperare la subnet da trustare nel `configuration.yaml` di Core, dovresti fare l'inspect non della rete `bridge` ma appunto di `domotica_default`

Es.
```
$ docker network inspect domotica_default | grep 172
                    "Subnet": "172.18.0.0/16",
                    "Gateway": "172.18.0.1"
```

Per quanto riguarda il "dove definire" la network da usare nel compose, come ti viene in mente di pensare di definirla nel `configuration.yaml` di core ???? Va "OVVIAMENTE" definita `nel docker-compose.yaml` (ti ho pure linkato la documentazione di compose...leggi prima di chiedere, per favore :) )

Tornando al tuo errore, fare le prove con l'app companion non ha alcun senso. Assicurati prima che la tua istanza sia correttamente raggiungibile facendo i test da browser (sia da rete interna che da internet).

**Reply by lnrd00** [reply]
Ti ringrazio per la correzione. La directory dove si trova il Compose l'ho chiamata `hass`, ed effettivamente la mia lista di network è così composta:
```
$ docker network ls
NETWORK ID     NAME           DRIVER    SCOPE
6b443b84af23   bridge         bridge    local
95fd5d34e188   hass_default   bridge    local
0d3ccd75d3d7   host           host      local
29396fe05aec   none           null      local
```
Ed utilizzando `docker network inspect hass_default` trovo la `"Subnet"` e il `"Gateway"` di mio interesse.

Per quanto riguarda dove definire la network, avevo letto la pagina a cui mi hai rimandato ma ciò non mi ha fermato dal commettere quell'errore sciocco, che mi ero reso conto di aver commesso già subito dopo avertelo scritto.
La domanda adesso è: la sezione di codice seguente 

```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - XXX.XXX.XXX.XXX
```
 deve essere "rimpiazzata" dal codice che mi hai mostrato te, è corretto?

Per quanto riguarda l'errore che riscontro ad accedere su HA, accedendo da rete interna con `http://INDIRIZZO_IP_DEL_RASPBERRY:8123` non ho alcun problema, mentre da rete esterna con `https://miodominio.duckdns.org:porta` la pagina rimane bianca e in caricamento. Sto continuando ad investigare per cercare da individuare il problema, ma per adesso niente da fare :(
Grazie ancora per il supporto!!

**Reply by snakuzzo** [reply]
> deve essere "rimpiazzata" dal codice che mi hai mostrato te, è corretto?

XXX.XXX.XXX.XXX deve essere sostituito con la `Subnet` che ti ritorna l'inspect della network del tuo compose (immagino `172.18.0.0/16`)

> mentre da rete esterna con https://miodominio.duckdns.org:porta la pagina rimane bianca e in caricamento

Eh ma se rimane in caricamento vuol dire che hai problemi di connettività. Devi assicurarti che il tuo IP pubblico sia uguale a quello del RECORD A relativo al tuo dominio e devi assicurarti che il port forwarding sia configurato correttamente sul router

**Reply by lnrd00** [reply]
>XXX.XXX.XXX.XXX deve essere sostituito con la Subnet che ti ritorna l'inspect della network del tuo compose (immagino 172.18.0.0/16)

mi chiedevo se, una volta definita staticamente la network all'interno del `compose.yaml`, allora avessi dovuto rimuovere la sezione
```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - XXX.XXX.XXX.XXX
```
all'interno del `configuration.yaml`, oppure mantenerla semplicemente aggiornandola con la subnet giusta al posto di XXX.XXX.XXX.XXX.


>Eh ma se rimane in caricamento vuol dire che hai problemi di connettività. Devi assicurarti che il tuo IP pubblico sia uguale a quello del RECORD A relativo al tuo dominio e devi assicurarti che il port forwarding sia configurato correttamente sul router

Era successo proprio questo!! 🥲Non mi ero reso conto che il mio IP pubblico si fosse aggiornato, adesso l'ho aggiornato sul sito web di DuckDNS e tutto funziona correttamente! 
P.S. C'è un modo per evitare di dover stare ad aggiornare l'IP pubblico, dunque di automatizzare tale processo? Non che sia niente di così fastidioso eh, del resto basta andare sul sito web ed aggiornarlo, però se sono fuori casa non mi sarebbe possibile individuare l'IP pubblico.

**Reply by snakuzzo** [reply]
> mi chiedevo se, una volta definita staticamente la network all'interno del compose.yaml, allora avessi dovuto rimuovere la sezione

ovviamente va lasciata!!! (però non puoi farmi queste domande, dovresti capire a cosa serve quella roba)

> oppure mantenerla semplicemente aggiornandola con la subnet giusta al posto di XXX.XXX.XXX.XXX.

va ovviamente mantenuta aggiornata con la subnet giusta, ma se la definisci nel compose (come ti ho detto più volte), questa sezione non la toccherai più

> C'è un modo per evitare di dover stare ad aggiornare l'IP pubblico, dunque di automatizzare tale processo?

ma non è che tu PUOI...tu DEVI tenerlo aggiornato costantemente!!!.
Per farlo puoi:
- farlo dal router direttamente, se te lo permette
- puoi configurare un altro container che lasci lì a fare questo (io faccio così per un'istanza di test) -> https://docs.linuxserver.io/images/docker-duckdns/#notice-regarding-automatic-detection
- puoi creare un package direttamente su core (io faccio così per un'altra istanza di test) -> https://indomus.it/guide/aggiornare-il-proprio-record-su-duckdns-tramite-automazione-home-assistant/

**Reply by lnrd00** [reply]
>puoi creare un package direttamente su core (io faccio così per un'altra istanza di test) -> https://indomus.it/guide/aggiornare-il-proprio-record-su-duckdns-tramite-automazione-home-assistant/

Ho seguito questa guida e credo di aver creato correttamente il package e tutto sembrerebbe funzionare, o meglio non c'è niente che non sta funzionando. Credo l'Odissea sia terminata 👍🏻
Grazie mille!!!
